client/client.go

```diff
@@ -233,9 +261,32 @@ func (c *httpClusterClient) reset(eps []string) error {
 		neps[i] = *u
 	}
 
-	c.endpoints = shuffleEndpoints(c.rand, neps)
-	// TODO: pin old endpoint if possible, and rebalance when new endpoint appears
-	c.pinned = 0
+	switch c.selectionMode {
+	case EndpointSelectionRandom:
+		c.endpoints = shuffleEndpoints(c.rand, neps)
+		c.pinned = 0
+	case EndpointSelectionPrioritizeLeader:
+		c.endpoints = neps
+		// TODO: should return ErrNoEndpoints in a case of getting leader fail?
+		lep, err := c.getLeaderEndpoint()
+		if err != nil {
+			return ErrNoEndpoints
+		}
+
+		lu, err := url.Parse(lep)
+		if err != nil {
+			return ErrNoEndpoints
+		}
+
+		for i := range c.endpoints {
+			if c.endpoints[i].String() == lu.String() {
```

Comment:
What if the leader isn't in the set of endpoints in a `Config` that gets passed to `client.New()`? Perhaps we should `Sync` during `New()` if `EndpointSelectionPrioritizeLeader`, to minimise the chance of that happening? and error out here if we don't find it?